placement/algo/algo.go

```diff
@@ -0,0 +1,482 @@
+	return totalLoad / numberOfHosts
+}
+
+func (ph placementHelper) getShardSet(shards []uint32) map[uint32]struct{} {
+	shardSet := make(map[uint32]struct{}, len(shards))
+	for _, shard := range shards {
+		shardSet[shard] = struct{}{}
+	}
+	return shardSet
+}
+
+func (ph placementHelper) canAssignRack(shard uint32, from *hostShards, toRack string) bool {
+	if from != nil {
+		if from.hostRack() == toRack {
+			return true
+		}
+	}
+	for host := range ph.shardToHostMap[shard] {
+		if host.hostRack() == toRack {
+			return false
+		}
+	}
+	return true
+}
+
+func (ph placementHelper) canAssignHost(shard uint32, from, to *hostShards) bool {
+	if to == nil {
+		return false
+	}
+	if _, exist := to.shardsSet[shard]; exist {
+		return false
+	}
+	return ph.canAssignRack(shard, from, to.hostRack())
+}
+
+func (ph placementHelper) assignShardToHost(shard uint32, to *hostShards) {
+	to.addShard(shard)
+
+	if _, exist := ph.shardToHostMap[shard]; !exist {
+		ph.shardToHostMap[shard] = make(map[*hostShards]struct{})
+	}
+	ph.shardToHostMap[shard][to] = struct{}{}
+}
+
+func (ph placementHelper) removeShardFromHost(shard uint32, from *hostShards) {
+	from.removeShard(shard)
+
+	delete(ph.shardToHostMap[shard], from)
+}
+
+func (ph placementHelper) placeShards(shards []uint32, from *hostShards) error {
+	shardSet := ph.getShardSet(shards)
+	var tried []*hostShards
+
+	if from != nil {
+		// prefer to distribute "some" of the load to other racks first
+		// because the load from a leaving host can always get assigned to a host on the same rack
+		ph.placeToRacksOtherThanOrigin(shardSet, from)
+	}
+
+	// if there are shards left to be assigned, distribute them evenly
+	for shard := range shardSet {
+		tried = tried[:0]
+		for ph.hostHeap.Len() > 0 {
+			tryHost := heap.Pop(ph.hostHeap).(*hostShards)
+			tried = append(tried, tryHost)
+			if ph.canAssignHost(shard, from, tryHost) {
+				ph.assignShardToHost(shard, tryHost)
+				for _, triedHost := range tried {
+					heap.Push(ph.hostHeap, triedHost)
+				}
+				break
+			}
+		}
+		if ph.hostHeap.Len() == 0 {
+			// this should only happen when RF > number of racks
+			return errNotEnoughRacks
+		}
+	}
+	return nil
+}
+
+// placeToRacksOtherThanOrigin move shards from a host to the rest of the cluster
+// the goal of this function is to assign "some" of the shards to the hosts in other racks
+func (ph placementHelper) placeToRacksOtherThanOrigin(shards map[uint32]struct{}, from *hostShards) {
+	var triedHosts []*hostShards
+	var sameRack []*hostShards
+outer:
+	for shard := range shards {
+		triedHosts = triedHosts[:0]
+		for ph.hostHeap.Len() > 0 {
+			tryHost := heap.Pop(ph.hostHeap).(*hostShards)
+			// do not place to same rack for now
+			if from != nil && tryHost.hostRack() == from.hostRack() {
+				sameRack = append(sameRack, tryHost)
+				continue
+			}
+			triedHosts = append(triedHosts, tryHost)
+			if ph.hostHeap.getTargetLoadForHost(tryHost.hostAddress())-len(tryHost.shardsSet) <= 0 {
```

Comment:
yeah there is a parameter when building the placementHelper, check newPlaceShardingHelper(ps placementSnapshot, targetRF int, hostCapacityAscending bool)